src/streams.c

```diff
@@ -81,32 +82,57 @@ static Int READ_COMMAND(Obj *evalResult)
- * cannot be opened, or executes all statements in the stream and returns a list of lists,
- * each entry of which reflects the result of the execution of one statement.
+ * This function reads all code from the stream and returns either `fail` if
+ * the stream cannot be opened, or executes all statements in the stream and
+ * returns a list of lists, each entry of which reflects the result of the
+ * execution of one statement.
  *
- * The results are returned as lists with at most three bound positions.
+ * The results are returned as lists of length three if printToOutput is
+ * false, or five, if printToOutput is true.
  *
- * If the first entry is `true`, then the second entry is bound to the result of
- * the statement if there was one, and unbound otherwise, and the third entry
- * reflects whether the statement ended in a dual semicolon,
- * if the first entry is `false', an error occurred during the execution of a statement,
- * and no other positions of the list are bound.
+ * If the first entry is `true`, then the second entry is bound to the result
+ * of the statement if there was one, and unbound otherwise, and the third
+ * entry reflects whether the statement ended in a dual semicolon, if the
+ * first entry is `false', an error occurred during the execution of a
+ * statement, and no other positions of the list are bound. If printToOutput
+ * is true, the fourth entry of the list is ViewObj applied to the result, if
+ * there was one, and if the statement did not end in a dual semicolon.
+ * Furthermore, if printToOutput is true, the fifth entry contains a string of
+ * all output printed while the corresponding statement was executed.
  *
- * This function is currently used in interactive tools such as the GAP Jupyter kernel to
- * execute cells and is likely to be replaced by a function that can read a single command
- * from a stream without losing the rest of its content.
+ * This function is currently used in interactive tools such as the GAP
+ * Jupyter kernel to execute cells and is likely to be replaced by a function
+ * that can read a single command from a stream without losing the rest of its
+ * content.
  */
-Obj FuncREAD_ALL_COMMANDS( Obj self, Obj stream, Obj echo )
+Obj FuncREAD_ALL_COMMANDS(Obj self, Obj instream, Obj echo, Obj printToOutput)
 {
     ExecStatus status;
     Int resultCount, resultCapacity;
     UInt dualSemicolon;
     Obj result, resultList;
     Obj evalResult;
+    Obj        outstream;
+    Obj        outstreamString;
+    Obj        outstreamStringCopy;
 
-    /* try to open the stream */
-    if (!OpenInputStream(stream, echo == True)) {
+    /* try to open the streams */
+    if (!OpenInputStream(instream, echo == True)) {
+        return Fail;
+    }
+
+
+    if (printToOutput == True) {
+        outstreamString = NEW_STRING(0);
+        outstream = DoOperation2Args(ValGVar(GVarName("OutputTextString")),
+                                     outstreamString, True);
```

Comment:
Hmm, is there a reason to create the stream here? Why not just let the caller pass in a stream object, and also accept e.g. `fail` to indicate the old behaviour? That would simplify the logic here quite a bit. But perhaps I am missing something?

Reply:
Oh I see, you need to create a custom string stream of course, and caputre its output after each single operation. Never mind then.